File: src/utils/dimensionParser.ts
import { AreaDisplayUnit, DimensionUnit } from '../types';

export class DimensionParser {
  static readonly METERS_TO_FEET = 3.280839895013123;
  static readonly METERS_TO_INCHES = 39.37007874015748;
  static readonly SQ_METERS_TO_SQ_FEET = 10.763910416709722;

  static feetInchesToMeters(feet: number, inches: number): number {
    const totalInches = feet * 12 + inches;
    return totalInches * 0.0254;
  }

  static meterCmToMeters(meters: number, cm: number): number {
    return meters + cm / 100.0;
  }

  static metersToFeet(meters: number): number {
    return meters * this.METERS_TO_FEET;
  }

  static metersToInches(meters: number): number {
    return meters * this.METERS_TO_INCHES;
  }

  static squareMetersToSquareFeet(squareMeters: number): number {
    return squareMeters * this.SQ_METERS_TO_SQ_FEET;
  }

  static squareFeetToSquareMeters(squareFeet: number): number {
    return squareFeet / this.SQ_METERS_TO_SQ_FEET;
  }

  static format(value: number, decimals: number = 2): string {
    return Number(value.toFixed(decimals)).toLocaleString('en-US', {
      minimumFractionDigits: 0,
      maximumFractionDigits: decimals,
    });
  }

  static formatFeetInches(meters: number): string {
    if (meters <= 0) return '0\' 0"';
    const totalInches = Math.round(meters * this.METERS_TO_INCHES);
    const feet = Math.floor(totalInches / 12);
    const inches = totalInches % 12;
    return `${feet}' ${inches}"`;
  }

  static formatMeterCm(meters: number): string {
    if (meters <= 0) return '0.00 m';
    return `${meters.toFixed(2)} m`;
  }

  static formatLength(meters: number, unit: AreaDisplayUnit = 'imperial'): string {
    if (meters <= 0) return '-';
    if (unit === 'metric') {
      return this.formatMeterCm(meters);
    }
    return this.formatFeetInches(meters);
  }

  static formatArea(squareFeet: number, unit: AreaDisplayUnit = 'imperial'): string {
    if (squareFeet <= 0 || isNaN(squareFeet)) return '0 sq ft';

    const sqMeters = this.squareFeetToSquareMeters(squareFeet);

    switch (unit) {
      case 'imperial':
        return `${this.format(squareFeet, 2)} sq ft`;
      case 'metric':
        return `${this.format(sqMeters, 2)} sq m`;
      case 'hybrid':
        return `${this.format(squareFeet, 2)} sq ft (${this.format(sqMeters, 2)} sq m)`;
    }
  }

  static getUnitLabel(unit: AreaDisplayUnit): string {
    switch (unit) {
      case 'imperial':
        return 'sq ft';
      case 'metric':
        return 'sq m';
      case 'hybrid':
        return 'sq ft / sq m';
    }
  }

  /**
   * Parses common dimension string formats into meters:
   * e.g. "12' 6\"", "12-6", "12.5 ft", "3.8 m", "3800 mm", "380 cm"
   */
  static parseDimensionToMeters(input: string): number {
    if (!input) return 0;
    const clean = input
      .trim()
      .replace(/[’‘`]/g, "'")
      .replace(/[“”]/g, '"')
      .toLowerCase();

    // Feet and inches: 10' 6", 10'6", 10'-6", 10-6, 10' - 6", 10 ft 6 in
    const feetInchesMatch = clean.match(/^(\d+(?:\.\d+)?)\s*(?:'|ft|feet)?\s*[-–—]?\s*(\d+(?:\.\d+)?)\s*(?:"|in|inch|inches)?$/);
    if (feetInchesMatch && (clean.includes("'") || clean.includes("ft") || clean.includes("-") || clean.includes("–") || clean.includes('"') || clean.includes("in") || clean.includes(" "))) {
      const feet = parseFloat(feetInchesMatch[1]);
      const inches = parseFloat(feetInchesMatch[2] || '0');
      return this.feetInchesToMeters(feet, inches);
    }

    // Feet only: 10' or 10 ft
    const feetOnlyMatch = clean.match(/^(\d+(?:\.\d+)?)\s*(?:'|ft|feet)$/);
    if (feetOnlyMatch) {
      const feet = parseFloat(feetOnlyMatch[1]);
      return feet * 0.3048;
    }

    // Inches only: 120" or 120 in
    const inchesOnlyMatch = clean.match(/^(\d+(?:\.\d+)?)\s*(?:"|in|inch|inches)$/);
    if (inchesOnlyMatch) {
      const inches = parseFloat(inchesOnlyMatch[1]);
      return inches * 0.0254;
    }

    // Millimeters: 3500 mm or 3500mm
    const mmMatch = clean.match(/^(\d+(?:\.\d+)?)\s*mm$/);
    if (mmMatch) {
      return parseFloat(mmMatch[1]) / 1000.0;
    }

    // Centimeters: 350 cm or 350cm
    const cmMatch = clean.match(/^(\d+(?:\.\d+)?)\s*cm$/);
    if (cmMatch) {
      return parseFloat(cmMatch[1]) / 100.0;
    }

    // Meters: 3.5 m or 3.5m or 3.5 meter
    const mMatch = clean.match(/^(\d+(?:\.\d+)?)\s*(?:m|meter|meters|metre|metres)$/);
    if (mMatch) {
      return parseFloat(mMatch[1]);
    }

    // Decimal number without unit
    const num = parseFloat(clean);
    if (!isNaN(num) && num > 0) {
      // Heuristic: if >= 500 it's probably millimeters (e.g. 3200)
      if (num >= 500) {
        return num / 1000.0;
      }
      // If between 20 and 500 it's probably centimeters (e.g. 350)
      if (num >= 50 && num < 500) {
        return num / 100.0;
      }
      // If <= 10 and has decimals, could be meters e.g. 3.25
      if (clean.includes('.') && num <= 15) {
        return num; // meters
      }
      // Otherwise default to feet (e.g. 10 or 12.5)
      return num * 0.3048;
    }

    return 0;
  }
}

File: android/app/src/main/kotlin/com/example/my_first_app/util/DimensionParser.kt
package com.example.my_first_app.util

import java.util.Locale
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.round

enum class AreaDisplayUnit {
    IMPERIAL,
    METRIC,
    HYBRID
}

enum class DimensionUnit {
    FEET_INCHES,
    METER_CM,
    DECIMAL_FEET
}

object DimensionParser {
    const val METERS_TO_FEET = 3.280839895013123
    const val METERS_TO_INCHES = 39.37007874015748
    const val SQ_METERS_TO_SQ_FEET = 10.763910416709722

    fun feetInchesToMeters(feet: Double, inches: Double): Double {
        val totalInches = feet * 12 + inches
        return totalInches * 0.0254
    }

    fun meterCmToMeters(meters: Double, cm: Double): Double {
        return meters + cm / 100.0
    }

    fun squareMetersToSquareFeet(squareMeters: Double): Double {
        return squareMeters * SQ_METERS_TO_SQ_FEET
    }

    fun squareFeetToSquareMeters(squareFeet: Double): Double {
        return squareFeet / SQ_METERS_TO_SQ_FEET
    }

    fun format(value: Double, decimals: Int = 2): String {
        return String.format(Locale.US, "%.${decimals}f", value)
            .replace(Regex("\\.0+$"), "")
            .replace(Regex("(\\.[0-9]*[1-9])0+$"), "$1")
    }

    fun formatFeetInches(meters: Double): String {
        if (meters <= 0) return "0' 0\""
        val totalInches = round(meters * METERS_TO_INCHES).toInt()
        val feet = totalInches / 12
        val inches = totalInches % 12
        return "$feet' $inches\""
    }

    fun formatMeterCm(meters: Double): String {
        if (meters <= 0) return "0.00 m"
        return String.format(Locale.US, "%.2f m", meters)
    }

    fun formatArea(squareFeet: Double, unit: AreaDisplayUnit = AreaDisplayUnit.IMPERIAL): String {
        if (squareFeet <= 0 || squareFeet.isNaN()) return "0 sq ft"
        val sqMeters = squareFeetToSquareMeters(squareFeet)
        return when (unit) {
            AreaDisplayUnit.IMPERIAL -> "${format(squareFeet, 2)} sq ft"
            AreaDisplayUnit.METRIC -> "${format(sqMeters, 2)} sq m"
            AreaDisplayUnit.HYBRID -> "${format(squareFeet, 2)} sq ft (${format(sqMeters, 2)} sq m)"
        }
    }

    fun parseDimensionToMeters(input: String): Double {
        if (input.isBlank()) return 0.0
        val clean = input.trim()
            .replace(Regex("[’‘`]"), "'")
            .replace(Regex("[“”]"), "\"")
            .lowercase()

        // Feet and inches: 10' 6", 10'6", 10-6, 10 ft 6 in
        val feetInchesRegex = Regex("^(\\d+(?:\\.\\d+)?)\\s*(?:'|ft|feet)?\\s*[-–—]?\\s*(\\d+(?:\\.\\d+)?)\\s*(?:\"|in|inch|inches)?$")
        val feetInchesMatch = feetInchesRegex.find(clean)
        if (feetInchesMatch != null && (clean.contains("'") || clean.contains("ft") || clean.contains("-") || clean.contains("–") || clean.contains("\"") || clean.contains("in") || clean.contains(" "))) {
            val feet = feetInchesMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            val inches = feetInchesMatch.groupValues[2].toDoubleOrNull() ?: 0.0
            return feetInchesToMeters(feet, inches)
        }

        // Feet only: 10' or 10 ft
        val feetOnlyRegex = Regex("^(\\d+(?:\\.\\d+)?)\\s*(?:'|ft|feet)$")
        val feetOnlyMatch = feetOnlyRegex.find(clean)
        if (feetOnlyMatch != null) {
            val feet = feetOnlyMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            return feet * 0.3048
        }

        // Inches only: 120" or 120 in
        val inchesOnlyRegex = Regex("^(\\d+(?:\\.\\d+)?)\\s*(?:\"|in|inch|inches)$")
        val inchesOnlyMatch = inchesOnlyRegex.find(clean)
        if (inchesOnlyMatch != null) {
            val inches = inchesOnlyMatch.groupValues[1].toDoubleOrNull() ?: 0.0
            return inches * 0.0254
        }

        // Millimeters: 3500 mm
        val mmRegex = Regex("^(\\d+(?:\\.\\d+)?)\\s*mm$")
        val mmMatch = mmRegex.find(clean)
        if (mmMatch != null) {
            return (mmMatch.groupValues[1].toDoubleOrNull() ?: 0.0) / 1000.0
        }

        // Centimeters: 350 cm
        val cmRegex = Regex("^(\\d+(?:\\.\\d+)?)\\s*cm$")
        val cmMatch = cmRegex.find(clean)
        if (cmMatch != null) {
            return (cmMatch.groupValues[1].toDoubleOrNull() ?: 0.0) / 100.0
        }

        // Meters: 3.5 m
        val mRegex = Regex("^(\\d+(?:\\.\\d+)?)\\s*(?:m|meter|meters|metre|metres)$")
        val mMatch = mRegex.find(clean)
        if (mMatch != null) {
            return mMatch.groupValues[1].toDoubleOrNull() ?: 0.0
        }

        // Decimal number without unit
        val num = clean.toDoubleOrNull()
        if (num != null && num > 0) {
            return when {
                num >= 500 -> num / 1000.0
                num >= 50 -> num / 100.0
                clean.contains(".") && num <= 15 -> num
                else -> num * 0.3048
            }
        }
        return 0.0
    }
}
